### DISCUSSION TITLE
fulltextsearch index - prevent reindexing

### DISCUSSION THREAD

**Discussion by ManOki** [nextcloud/all-in-one#1694] [Ideas] [answered]
In case of migrating large setups with a lot of data, there will (imho) always be some errors while indexing fulltextsearch. This results into a complete laborious re-index every time the container is start up (without nextcloud available for the users).

Workaround: create marker file [$NEXTCLOUD_DATA_DIR/fts-index.done](https://github.com/nextcloud/all-in-one/blob/0588a3391cb0542f99c9adafc651cf38fd738916/Containers/nextcloud/entrypoint.sh#L591) in nextcloud container.

See #1692

**Comment by szaimen** [comment]
Thanks for creating this discussion!

In my opinion should the ultimate goal be to make the `fulltextsearch:index` as solid as possible. But this does not work without reporting errors to the [upstream project](https://github.com/nextcloud/fulltextsearch/issues). Simply always ignoring issues is not something I would encourage people to do.

---

So my idea would be to create a new wiki article [here](https://github.com/nextcloud/all-in-one/discussions/new?category=wiki) with what to do if you run into issues with the index:
1. report the error with logs upstream if there shouldn't be already an open one about this specific issue
2. create the marker file.

We could then directly link from here to the wiki article: https://github.com/nextcloud/all-in-one/blob/0588a3391cb0542f99c9adafc651cf38fd738916/Containers/nextcloud/entrypoint.sh#L598

WDYT? :)

---

If you think this is a good idea, can you please create the wiki aricle? I can then add the link into the code :)

**Reply by ManOki** [reply]
> In my opinion should the ultimate goal be to make the `fulltextsearch:index` as solid as possible. But this does not work without reporting errors to the [upstream project](https://github.com/nextcloud/fulltextsearch/issues). Simply always ignoring issues is not something I would encourage people to do.

I agree _but_ the problem is that `fulltextsearch:index` is [interactive](https://github.com/nextcloud/fulltextsearch/wiki/Commands#fulltextsearchindex) (on the console). This leads to a bunch of problems anyway: 
- `docker logs` is basically spammed (for minor output reduce add parameter `--no-interaction --no-readline`)
- you can't interact with entrypoint.sh to get the real errors/files
- there is no separate error log file (at least I haven't found any) which could be further processed

I see output like this:

```
┌─ Indexing  ────
│ Action: generateIndexFiles
│ Provider: Files                Account: user
│ Document:
│ Info: /user/files/filepath
│ Title:
│ Content size:
│ Chunk:      2/406
│ Progress:      0/532
└──
┌─ Results ────
│ Result:      0/0
│ Index:
│ Status:
│ Message:
│
│
└──
┌─ Errors ────
│ Error:      5/5
│ Index: files:4304344
│ Exception: Elasticsearch\Common\Exceptions\BadRequest400Exception
│ Message: Expected string 'obj' but missed at character 'o' at offset 167
│
│
└──
## x:first result ## c/v:prec/next result ## b:last result
## f:first error ## h/j:prec/next error ## d:delete error ## l:last error
## q:quit ## p:pause
```

So I have a lot of files and therefore a log of logs but _only_ 5 errors. Fulltextsearch somehow saves them as 5 is shown again right after start of re-index. FYI: You could reset them using `occ fulltextsearch:index "{\"errors\": \"reset\"}"`. For me there is no easy way to list/show those errors 😐.

> We could then directly link from here to the wiki article:
> https://github.com/nextcloud/all-in-one/blob/0588a3391cb0542f99c9adafc651cf38fd738916/Containers/nextcloud/entrypoint.sh#L598

This could be missed the first time due to extensive logs and in worst case users will have to reindex several times before getting there.

So a non-interactive mode might be a feature request to nextcloud/fulltextsearch.

**Reply by szaimen** [reply]
Good point regarding the interactive problem. Looks like we might be able to work around this with this undocumented option: https://github.com/nextcloud/fulltextsearch/blob/f333408dfb4443e9abd2a518c4e7cdce76eb6608/lib/Command/Index.php#L204

**Reply by szaimen** [reply]
WIP here: https://github.com/nextcloud/all-in-one/pull/1695

**Reply by szaimen** [reply]
Now that this is solved, do you mind creating the wiki article? :)

**Comment by ManOki** [comment, marked as answer]
I submitted a feature request, see [fulltextsearch #731](https://github.com/nextcloud/fulltextsearch/issues/731).

**Reply by szaimen** [reply]
Great, thanks!

**Reply by szaimen** [reply]
I've created https://github.com/nextcloud/all-in-one/discussions/1709 now

**Reply by szaimen** [reply]
This is now released with v4.2.0 Beta. Testing and feedback is welcome! See https://github.com/nextcloud/all-in-one#how-to-switch-the-channel
